File: archive/cpp/2021/oct/invert-binary-tree.cpp
#include "common.h"

class Solution
{
public:
    TreeNode *invertTree(TreeNode *root)
    {
        using std::swap;
        if (root)
        {
            invertTree(root->left);
            invertTree(root->right);
            std::swap(root->left, root->right);
        }
        return root;
    }
};

File: src/tree/invert_binary_tree.rs
use crate::{Solution, TreeNode};
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn invert_tree(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        match root {
            Some(root) => {
                let l = root.borrow().left.clone();
                let r = root.borrow().right.clone();

                root.borrow_mut().left = Self::invert_tree(r);
                root.borrow_mut().right = Self::invert_tree(l);

                Some(root)
            }
            None => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test() {
        println!("{}", Solution::default());
    }
}
